sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriterOptions.cs

```diff
@@ -33,4 +38,115 @@ public ModelReaderWriterOptions (string format)
+    }
+
+    /// <summary>
+    /// Gets the model that is currently being proxied.
+    /// </summary>
+    /// <remarks>
+    /// Whenever a proxy is used for a given type, the model being proxied is set here.
+    /// The value does not need to be cleared after setting it will be overwritten on the next use.
+    /// </remarks>
+    public object? ProxiedModel { get; private set; }
+
+    /// <summary>
+    /// Gets the <see cref="IPersistableModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
+    /// </summary>
+    /// <param name="proxy"> The <see cref="IPersistableModel{T}"/> proxy if one exists. </param>
+    /// <returns> True if a proxy was found; otherwise, false. </returns>
+    public bool TryGetProxy<T>([NotNullWhen(true)] out IPersistableModel<T>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(typeof(T), out object? result))
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = (IPersistableModel<T>)result;
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="IJsonModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
+    /// </summary>
+    /// <param name="proxy"> The <see cref="IJsonModel{T}"/> proxy if one exists. </param>
+    /// <returns> True if a proxy was found; otherwise, false. </returns>
+    public bool TryGetProxy<T>([NotNullWhen(true)] out IJsonModel<T>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(typeof(T), out object? result) || result is not IJsonModel<T> jsonResult)
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = jsonResult;
+        return true;
+    }
+
+    internal bool TryGetProxy(Type modelType, [NotNullWhen(true)] out IJsonModel<object>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(modelType, out object? result) || result is not IJsonModel<object> jsonResult)
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = jsonResult;
+        return true;
+    }
+
+    internal bool TryGetProxy(Type modelType, [NotNullWhen(true)] out IPersistableModel<object>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(modelType, out object? result))
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = (IPersistableModel<object>)result;
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="IPersistableModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
+    /// </summary>
+    /// <param name="model"> The <see cref="IPersistableModel{T}"/> model to proxy. </param>
+    /// <returns> The <see cref="IPersistableModel{T}"/> proxy to use. </returns>
+    public IPersistableModel<T> ResolveProxy<T>(IPersistableModel<T> model)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(model.GetType(), out object? result))
```

Comment:
It might be helpful for future readers/maintainers of this code to add a comment that explains that this checks whether a proxy has been registered for type T, and maybe also that this implements an exact type look-up, and not a proxy for subtypes of T.

Reply:
I think the line of code is clear that it is checking if a proxy exists, but added a line at the top of the class stating all proxies are exact matches. since this is the case for all methods not just this one.